### PR TITLE
[FIX] Load compcor related confounds generated from fmriprep 21.x series

### LINKED ISSUE #3268: nilearn.interfaces.fmriprep.load_confounds not able to load compcor regressors
When using fmriprep version > 21.0.0 and a confound strategy that attempts to load 'WM' or 'CSF' acompcor regressors, no regressors are loaded. 

The reason seems to be that the naming convention on fmriprep's side was changed in fmriprep version 21.0.0 (Dec 2021, see [fmriprep changelog](https://fmriprep.org/en/stable/changes.html#december-14-2021) "FIX: Split aCompCor into aCompCor, cCompCor, wCompCor (# 2523)"). The regressors for aCompCor on the anatomical masks 'WM' and 'CSF' have been renamed. 

Nilearn version:
0.9.2.dev
### Expected behavior
Load acompcor regressors computed on separate anatomical masks

### Actual behavior
return empty list

### Steps and code to reproduce bug

```python
confounds, _=fmriprep.load_confounds(
            img_path, strategy=['compcor','high_pass'],compcor='anat_separated')
# confounds does not contain acompcor regressors although they exist
```

Code that needs to be changed (?)
https://github.com/nilearn/nilearn/blob/a2bca9df73bd397894fca7bc193519904767bf79/nilearn/interfaces/fmriprep/load_confounds_compcor.py#L4-L6
````
prefix_compcor = {"temporal_anat": ["t", "a"],
                  "temporal": ["t"],
                  "anat": ["a","c","w"]}
````

### PR DESCRIPTION
**Pull request by htwangtw**
Closes #3268. 

- refactor the test file loader to allow adding new major release 
- add test data for 21.x.x and make sure the correct files are loaded
- apply compcor loading suggestions from @askieslinger and iterate on it
- simply `_check_compcor_method` by using all five options as prefix checking keys
- add tests to `test_load_confounds_strategy.py::test_strategy_compcor` and `test_load_confounds.py::test_n_compcor` to make sure the correct components are loaded for 21.x and older

### CODE REVIEW

File: nilearn/interfaces/fmriprep/tests/test_load_confounds_strategy.py [hunk 1]
```diff
@@ -120,6 +121,20 @@ def test_strategy_compcor(tmp_path):
         "a_comp_cor_57" not in compcor_col_str_anat
     )  # this one comes from the white matter mask
 
+    file_nii, _ = create_tmp_filepath(
```
Comment (NicolasGensollen) [inline nilearn/interfaces/fmriprep/tests/test_load_confounds_strategy.py:124]: I think you can factorize this with a fixture. Maybe something like this:

```python
@pytest.mark.parametrize("fmriprep_version", ["1.4.x", "21.x.x"])
def test_strategy_compcor(tmp_path):
     """Check user specified input for compcor strategy."""
     file_nii, _ = create_tmp_filepath(
         tmp_path, image_type="regular", copy_confounds=True, copy_json=True,
         fmriprep_version=fmriprep_version
     )
     confounds, _ = load_confounds_strategy(
         file_nii, denoise_strategy="compcor"
    )
    compcor_col_str_anat = "".join(confounds.columns)
    assert "t_comp_cor_" not in compcor_col_str_anat
    expected = "a_comp_cor_57" if fmriprep_version == "1.4.x" else "w_comp_cor_00"
    assert expected not in compcor_col_str_anat  # this one comes from the white matter mask
```

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @htwangtw !
I made a quick first pass and it LGTM. I just have a small suggestion to refactor a test.
Also, could you add a whatsnew ?

**Review by bthirion — COMMENTED** [review]
I think that the PR is fine, but this leads us exactly where we didn't want to go: have an increasingly complex infrastructure to be compatible with fmriprep.
I would be in favor of deprecating old fmriprep versions asap.

File: doc/changes/latest.rst [hunk 2]
```diff
@@ -24,6 +24,9 @@ Fixes
 - Function ``nilearn.glm.second_level.second_level._check_second_level_input`` now raises an error when ``flm_object`` argument is ``False`` and ``second_level_input`` is a list of :class:`~glm.first_level.FirstLevelModel` (:gh:`3283` by `Matthieu Joulot`_).
 - Function :func:`~image.resample_img` now warns the user if the provided image has an ``sform`` code equal to 0 or None  (:gh:`3284` by `Matthieu Joulot`_).
 - Fix usage of ``scipy.stats.gamma.pdf`` in ``_gamma_difference_hrf`` function under ``nilearn/glm/first_level/hemodynamic_models.py``, which resulted in slight distortion of HRF (:gh:`3297` by `Kun CHEN`_).
+- Update the `CompCor`` strategy in :func:`~interfaces.fmriprep.load_confounds`
```
Comment (NicolasGensollen) [inline doc/changes/latest.rst:27]:
```suggestion
- Update the ``CompCor`` strategy in :func:`~interfaces.fmriprep.load_confounds` 
```

**Review by NicolasGensollen — COMMENTED** [review]
Just a small typo

**Comment by htwangtw** [comment]
@bthirion @NicolasGensollen Let me know if there's more we want to discuss about this PR!

**Review by bthirion — APPROVED** [review]
I don't see any outstanding issue. But take my review with a grain of salt, I'm not familar at all with this part of the codebase.
Thx !

File: nilearn/interfaces/fmriprep/tests/utils.py [hunk 3]
```diff
@@ -88,9 +90,10 @@ def create_tmp_filepath(
     return tmp_img, tmp_conf
 
 
-def get_leagal_confound(non_steady_state=True):
+def get_leagal_confound(non_steady_state=True, fmriprep_version="1.4.x"):
```
Comment (tsalo) [inline nilearn/interfaces/fmriprep/tests/utils.py:93]: Should this be `get_legal_confound`?

File: nilearn/interfaces/fmriprep/load_confounds_compcor.py [hunk 4]
```diff
@@ -42,22 +44,17 @@ def _select_compcor(compcor_cols, n_compcor):
     # only select if not "all", or less components are requested than there
     # actually is
     if (n_compcor != "all") and (n_compcor < len(compcor_cols)):
-        compcor_cols = compcor_cols[0:n_compcor]
+        compcor_cols = compcor_cols[:n_compcor]
```
Comment (tsalo) [inline nilearn/interfaces/fmriprep/load_confounds_compcor.py:47]: @htwangtw would it make sense to add a check around here? Something like a warning if no CompCor regressors are found?

Comment (htwangtw) [reply nilearn/interfaces/fmriprep/load_confounds_compcor.py:47]: Good point - I have yet encountered this case but is it possible with how compcor is done?
I know it's possible for fMRIPrep to output confounds with no `cosineXX` regresssors

File: nilearn/interfaces/fmriprep/tests/utils.py (hunk 3 above)
Comment (htwangtw) [reply nilearn/interfaces/fmriprep/tests/utils.py]: Fixed now :)

File: nilearn/interfaces/fmriprep/load_confounds_compcor.py (hunk 4 above)
Comment (tsalo) [reply nilearn/interfaces/fmriprep/load_confounds_compcor.py:47]: I hadn't really considered that. I don't know if fMRIPrep would ever not write out compcor regressors, short of a failure within the workflow. I was more thinking that, if fMRIPrep changes things again, having warnings in these functions would make it easier for users to tell something's gone wrong.

Comment (htwangtw) [reply nilearn/interfaces/fmriprep/load_confounds_compcor.py:47]: Right now the behaviour is the function will just fail, and a check is implemented in `_load_single_confounds_file`, so we don't need to write things in each function

https://github.com/nilearn/nilearn/blob/ec57e566455b3d929480fc2ce19e151e16673c6e/nilearn/interfaces/fmriprep/load_confounds.py#L315

The only variable that could be missing when fMRIPrep can run correctly is `cosineXX`, when the timeseries is too short. 

This deserves it's own issue

Comment (tsalo) [reply nilearn/interfaces/fmriprep/load_confounds_compcor.py:47]: Sorry for the delay in responding. I don't think `_check_error` catches the missing CompCor regressors. I tried running `load_confounds` on an fMRIPrepped file from 22.0.0 with the following kwargs:

```
{'strategy': ['motion', 'high_pass', 'compcor'],
 'motion': 'derivatives',
 'compcor': 'anat_separated',
 'n_compcor': 5}
```

It successfully ran, but the only regressors it returned were the cosine terms, motion parameters, and the motion derivatives. I wouldn't expect 22.x acompcor confounds to be loaded correctly per #3268, but I would have expected an error.

Comment (htwangtw) [reply nilearn/interfaces/fmriprep/load_confounds_compcor.py:47]: No worries! I really appreciate that you took time to look into this!
I will have more time to work on this next week..

**Comment by Remi-Gau** [comment]
@htwangtw 
what's the status on this PR? The number of merge conflicts is starting to worry me...

**Comment by htwangtw** [comment]
> @htwangtw what's the status on this PR? The number of merge conflicts is starting to worry me...

It's pretty much ready and then I forgot to remind people to review it.... I can fix it up this week.

**Comment by Remi-Gau** [comment]
@htwangtw 
You said you were drowning in notifications and similar things.
Do you want me to take this one of your plate?

File: nilearn/interfaces/fmriprep/load_confounds_compcor.py (hunk 4 above)
Comment (Remi-Gau) [reply nilearn/interfaces/fmriprep/load_confounds_compcor.py]: checked on a 23.0.2 dataset

```python
from nilearn.interfaces.fmriprep import load_confounds

nii = "/home/remi/datalad/datasets.datalad.org/openneuro-derivatives/ds003768-fmriprep/sub-01/func/sub-01_task-rest_run-1_desc-preproc_bold.nii.gz"

confounds_out, sample_mask_out = load_confounds(
    nii,
    strategy=["motion", "high_pass", "compcor"],
    motion="derivatives",
    compcor="anat_separated",
    n_compcor=5,
)

print(sorted(confounds_out.columns))
```

and got this:

```python
['c_comp_cor_00', 'c_comp_cor_01', 'c_comp_cor_02', 'c_comp_cor_03', 'c_comp_cor_04', 'cosine00', 'cosine01', 'cosine02', 'cosine03', 'cosine04', 'cosine05', 'cosine06', 'cosine07', 'rot_x', 'rot_x_derivative1', 'rot_y', 'rot_y_derivative1', 'rot_z', 'rot_z_derivative1', 'trans_x', 'trans_x_derivative1', 'trans_y', 'trans_y_derivative1', 'trans_z', 'trans_z_derivative1', 'w_comp_cor_00', 'w_comp_cor_01', 'w_comp_cor_02', 'w_comp_cor_03', 'w_comp_cor_04']
```

AFAICT this test also catches missing regressors:

https://github.com/nilearn/nilearn/blob/68cfa06023edd4111aea681b6a157cc1cb2ac5e0/nilearn/interfaces/fmriprep/tests/test_load_confounds_components.py#L30

**Comment by Remi-Gau** [comment]
@htwangtw 

- reduced the number of rows in the new confound files to 30 (like the one that was already there)
- fixed merge conflicts
- parametrized several tests to be run on both fmriprep data versions
- refactored some tests a bit

**Comment by htwangtw** [comment]
Thanks a bunch @Remi-Gau !

**Review by ymzayek — APPROVED** [review]
LGTM thanks!

**Comment by Remi-Gau** [comment]
OK will merge this.

@tsalo 
if your comment from above still stands, let's open an issue and tackle that in another PR.

**Pull request merged**